### PR TITLE
Support multiple arguments in PipelineStage forward

### PR DESCRIPTION
**Pull request by H-Huang**
Currently the `PipelineStage` API wraps around an `nn.module` but does not support `nn.module`s which have multiple arguments in their `forward()`. This updates `PipelineStage` so that the forward may take an arbitrary number of arguments (tensors only) and corresponding p2p ops (`get_fwd_recv_ops`, `get_bwd_send_ops`, etc.) which are used in the pipeline schedules are still valid.

Tested on all the current schedules

### CODE REVIEW

File: pippy/PipelineSchedule.py [hunk 1]
```diff
@@ -19,10 +19,8 @@
 
 class PipelineStage(ABC, nn.Module):
     @abstractmethod
-    def forward(self, microbatch):
+    def forward(self, args: List[torch.tensor]) -> torch.tensor:
```
Comment (fegin) [inline pippy/PipelineSchedule.py:22]: Is it possible to support `*args` and `**kwargs` for more general use cases?

Comment (H-Huang) [reply pippy/PipelineSchedule.py:22]: @fegin Yes, definitely possible. Right now the component that is actually calling `forward()` is the `PipelineSchedule`, for example for gpipe [here](https://github.com/pytorch/PiPPy/blob/main/pippy/PipelineSchedule.py#L247). As you can see the forward is taking in a single microbatch. So how the `forward()` of PipelineStage looks is going to depend on how/if we want to enforce the structure of the microbatches to be. This is still TBD, but I expect I will make further changes after this.

File: pippy/PipelineSchedule.py [hunk 2]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
-        ), "Waiting for backward input without enqueueing one"
-        self.bwd_recv_queue.wait()
-        self.bwd_recv_queue = None
-        return self.fwd_output_grads
+    def get_bwd_send_ops(self) -> List[dist.P2POp]:
+        if self.is_first_stage:
+            return []
+        for fwd_input in self.fwd_inputs:
+            logger.info(f"{fwd_input.grad=}")
+            assert fwd_input.grad is not None, "grad must be valid"
```
Comment (kwen2501) [inline pippy/PipelineSchedule.py:221]: Can grad be None in some cases? e.g. some inputs do not require grad. We may want to loosen this later?

File: pippy/PipelineSchedule.py [hunk 3]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
-        return [dist.P2POp(dist.irecv, self.fwd_input, self.prev_stage)]
+        return [
+            dist.P2POp(dist.irecv, fwd_input, self.prev_stage)
+            for fwd_input in self.fwd_inputs
+        ]
 
     def get_fwd_send_ops(self) -> List[dist.P2POp]:
+        assert (
+            self.fwd_outputs is not None
+        ), "forward() must be called before get_fwd_send_ops"
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:174]: Do we see this in practice? If a module forward is None then it's not being used / shouldn't be pipelined right? It should be easy to support, but not sure when this case would come up.

File: pippy/PipelineSchedule.py [hunk 4]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
+
+        return self.fwd_inputs
 
     def compute_loss(self):
-        if self.fwd_output is None:
+        if self.fwd_outputs is None:
             raise RuntimeError("forward() must be called before compute_loss()")
         # TODO: use a real loss function passed in
-        return self.fwd_output.mean()
+        return self.fwd_outputs[0].mean()
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:267]: the `[0]` is just the first return value of the nn.module. This function is just a placeholder and needs to be updated so the users can actually pass in a loss function

File: pippy/PipelineSchedule.py [hunk 5]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
-            self.fwd_input = microbatch
+            self.fwd_inputs = args
 
         # this is needed when we access the gradients for this in backward()
-        self.fwd_input.requires_grad = True
-        self.fwd_input.retain_grad()
+        # logger.info(f"len fwd inputs {len(self.fwd_inputs)}")
+        # logger.info(f"{self.fwd_inputs=}")
+        for tensor in self.fwd_inputs:
+            tensor.requires_grad = True
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:191]: for first stage we probably don't need gradients for its inputs. In all the other stages we need the gradients given the input is used during computation.

File: pippy/PipelineSchedule.py (hunk 2 above)
Comment (H-Huang) [reply pippy/PipelineSchedule.py:221]: Yes we can loosen this. Also right now even if the inputs do not require grads we send backwards a tensor of all zeros for it's gradients which is wasteful.

File: pippy/PipelineSchedule.py [hunk 6]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
         output_for_backward = (
-            self.compute_loss() if self.is_last_stage else self.fwd_output
+            self.compute_loss() if self.is_last_stage else self.fwd_outputs
         )
 
         # we store a ref to the input/output pair for this forward to be later used by the corresponding backward
         self.fwd_outputs_for_backward.append(
-            (self.fwd_input, output_for_backward)
+            (self.fwd_inputs, output_for_backward)
         )
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:204]: makes sense 2 me!

File: pippy/PipelineSchedule.py [hunk 7]
```diff
@@ -141,83 +163,108 @@ def init_p2p_neighbors(self):
...
-            torch.autograd.backward(
-                fwd_outputs, self.fwd_output_grads, retain_graph=True
+            gradients = torch.autograd.grad(
+                outputs=fwd_outputs,
+                inputs=self.fwd_inputs,
+                grad_outputs=self.fwd_output_grads,
+                retain_graph=True,
+                allow_unused=True,
+                materialize_grads=True,
             )
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:255]: let me try, i need to look into `torch.autograd.grad` more closely

File: pippy/PipelineSchedule.py (hunk 2 above)
Comment (kwen2501) [reply pippy/PipelineSchedule.py:219]: I see, thanks. Do you mind putting a `# HACK` mark to describe the above presumption? Just for information purpose later.

**Review by kwen2501 — APPROVED** [review]
Added responses to some comments. But the PR is in landable state (would be nice to add some comments in the code to give a heads-up for some presumptions)

File: pippy/PipelineSchedule.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:265]: hmm i see, good point. Is there ever a case when users would want to get the loss for an individual microbatch? Or in practice do we always see users reducing the loss over all of the microbatches?

Regardless I agree with you, I can think about and add an API for that.
